panic: runtime error: index out of range [1] with length 1 -- when background containers running

I am pretty new to docker, but I don't think I should be seeing go errors as an end user, so I am notifying the community of this bug. 
I updated Docker, to version: v2.0.0-rc.1, and then started getting:  panic: runtime error: index out of range [1] with length 1, followed by some "go" errors when running docker-compose up. After searching on here, I found similar error messages with using 0 replicas docker/compose-cli#1935, but the error is different (1 instead of 0), and I wasn't using replicas -- unless I am missing something. 
Ultimately I ended up using 0 replicas as a workaround to get a proper error message, which was that I had background containers running.

My guess is that something similar to docker/compose-cli#1935 can be implemented to prevent these errors as well.



**Steps to reproduce the issue:**
1. Updated to version: v2.0.0-rc.1
2. Had containers running in the background
3. ran "docker-compose up" with this docker-compose.yml file:
services:
    api:
      container_name: api
      build:
        context: .
        dockerfile: api/Dockerfile.api
      image: mutcompute-api
      ports:
        - "5000:5000"
      volumes:
        - ./api:/app/
      environment: 
        - MAIL_SERVER=email
        - MAIL_PORT=8025
      command: gunicorn -b 0.0.0.0:5000 api:app
    client:
      container_name: client
      build:
        context: .
        dockerfile: client/Dockerfile.client
      image: mutcompute-client
      ports:
        - "3000:3000"
      volumes:
        - ./client:/app/
    email:
      container_name: email
      build:
        context: .
        dockerfile: api/Dockerfile.api
      image: mutcompute-api
      ports:
        - "8025:8025"
      command: python -m smtpd -n -c DebuggingServer 0.0.0.0:8025 

**Describe the results you received:**

panic: runtime error: index out of range [1] with length 1

goroutine 37 [running]:
github.com/docker/compose-cli/pkg/compose.(*convergence).ensureService(0xc00027ff20, 0x2120ba8, 0xc000454450, 0xc000240f00, 0xc0001ce7c8, 0x3, 0x0, 0x0, 0x0, 0xc000240a00, ...)
        github.com/docker/compose-cli/pkg/compose/convergence.go:222 +0x11f1
github.com/docker/compose-cli/pkg/compose.(*convergence).apply.func1(0x2120ba8, 0xc000454450, 0xc0001ce7c8, 0x3, 0x0, 0x0)
        github.com/docker/compose-cli/pkg/compose/convergence.go:99 +0x21f
github.com/docker/compose-cli/pkg/compose.run.func1(0x0, 0x0)
        github.com/docker/compose-cli/pkg/compose/dependencies.go:102 +0xa3
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000590540, 0xc000626180)
        golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x66

**Describe the results you expected:**

Error response from daemon: driver failed programming external connectivity on endpoint <...> Bind for 0.0.0.0:5000 failed: port is already allocated

**Additional information you deem important (e.g. issue happens only occasionally):**
Heres my SO post:
https://stackoverflow.com/questions/68900909/docker-compose-error-panic-runtime-error-index-out-of-range-1-with-length?noredirect=1#comment121769539_68900909

And an exerpt:
How I  fixed it:

I switched to a previous version without db service and was prompted to run docker-compose --remove -orphans, which didn't really do much. After searching through GitHub issues, and seeing most of the issues were related to replicas, I decided what the heck and added a replica section to the api service. While this error had nothing to do with replicas, I was able to get a different error:

WARNING: The "api" service is using the custom container name "api". Docker requires each container to have a unique name. Remove the custom name to scale the service.

for this error, I just commented out all the container_name sections in each service, and got a new error:

Error response from daemon: driver failed programming external connectivity on endpoint <...> Bind for 0.0.0.0:5000 failed: port is already allocated

Boom! this was the real issue. Searched SO for this, and found out I just needed to run docker-compose down

When I switched back to the branch with the db service, all was working!

**Output of `docker-compose --version`:**
```
Docker Compose version v2.0.0-rc.1
```

**Output of `docker version`:**

```
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:55:20 2021
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:10 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker context show`:**  
You can also run `docker context inspect context-name` to give us more details but don't forget to remove sensitive content.

```
default
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  compose: Docker Compose (Docker Inc., v2.0.0-rc.1)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 23
  Running: 0
  Paused: 0
  Stopped: 23
 Images: 30
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.47-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 1.939GiB
 Name: docker-desktop
 ID: AJ7X:ZSOB:YBIX:AXRV:UMDE:5E3J:UJK4:DHD7:743K:M4TV:SCYI:QXBS
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 42
  Goroutines: 43
  System Time: 2021-08-30T23:36:08.470272Z
  EventsListeners: 3
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS ECS, Azure ACI, local, etc.):**
Local environment

--- a/pkg/compose/convergence.go
+++ b/pkg/compose/convergence.go
@@ -171,7 +171,7 @@ func (c *convergence) ensureService(ctx context.Context, project *types.Project,
 	eg, _ := errgroup.WithContext(ctx)
 
 	for i, container := range containers {
-		if i > expected {
+		if i >= expected {
 			// Scale Down
 			container := container
 			eg.Go(func() error {

